app/src/main/java/ch/eureka/eurekapp/model/downloads/DownloadService.kt

```
@@ -0,0 +1,39 @@
+// Portions of this code were generated with the help of Grok.
+package ch.eureka.eurekapp.model.downloads
+
+import android.content.Context
+import android.net.Uri
+import androidx.core.content.FileProvider
+import java.io.File
+import java.io.FileOutputStream
+import java.net.URL
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class DownloadService(
+    private val context: Context,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+
+  suspend fun downloadFile(url: String, fileName: String): Uri? =
+      withContext(dispatcher) {
+        try {
+          val downloadsDir = File(context.cacheDir, "downloads")
+          if (!downloadsDir.exists()) {
+            downloadsDir.mkdirs()
+          }
+
+          val file = File(downloadsDir, fileName)
+
+          URL(url).openStream().use { input ->
+            FileOutputStream(file).use { output -> input.copyTo(output) }
+          }
+
+          FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
+        } catch (e: Exception) {
+          e.printStackTrace()
+          null
```

Why not return a Result \<Uri>, here the code fails silently